---
title: "Tidybiology Final Project"
author: "Your name here"
date: Sys.date()
---
## Load libraries
```{r setup}
library(tidyverse)
library(here)
library(glue)
library(ggrepel)
library(ggforce)

#uncomment (or select) to clear environment; good for provenance
rm(list=ls()) 
```

## Import data
I first imported genome summary data, which was originally scraped from [a Wikipedia entry about the human genome](https://en.wikipedia.org/wiki/Human_genome). 
```{r scrape, eval=FALSE, include=FALSE}
#alt code chunk to scrape directly
library(rvest)
library(janitor)

url <- read_html("https://en.wikipedia.org/wiki/Human_genome")

genome <- url %>% 
  html_nodes(xpath = '//*[@id="mw-content-text"]/div/table[2]') %>% #got xpath from inspect element, copy xpath
  html_table()

genome <- genome[[1]]
genome <- as_tibble(genome, .name_repair = "minimal") %>% 
  clean_names()
genome <- genome %>% 
  filter(!str_detect(chromosome, "mtDNA")) %>% #remove mitochondrial dna
  filter(!str_detect(chromosome, "total")) %>% 
  select(-links) %>% 
  select(-cumulative_percent)

genome$basepairs <- str_remove_all(genome$basepairs, "\\,")
genome$basepairs <- as.numeric(genome$basepairs)

genome$variations <- str_remove_all(genome$variations, "\\,")
genome$variations <- as.numeric(genome$variations)

genome$centromereposition_mbp <- as.numeric(genome$centromereposition_mbp)

chromosome_levels <- c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "X", "Y")
genome$chromosome <- fct_relevel(genome$chromosome, chromosome_levels)
write_csv(genome, path = here::here("genome.csv"))
```

```{r import}
genome <- read_csv(here::here("5_markdown", "genome.csv"), col_names = TRUE)
#cleaning step, required to relevel chromosomes in proper order
chromosome_levels <- c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "X", "Y")
genome$chromosome <- fct_relevel(genome$chromosome, chromosome_levels)

#add some new variables
genome <- genome %>% 
  mutate(percent_var = variations/basepairs) %>% #this adds a new column called percent_var to calculate the number of variations/basepairs
  mutate(xy = if_else(str_detect(chromosome, "X") | str_detect(chromosome, "Y"), TRUE, FALSE))  #adds a boolean for whether this is an XY chromosome or not to call them out in the graph
```

## Take a `glimpse` into the data in the dataframe
This table has 14 variables and 24 observations (22 numbered chromosomes + X and Y).
```{r glimpse}
glimpse(genome)
```
## EDA
I began to explore the relationship between number of protien codeing genes and number of base pairs. Perhaps not surprising, generally longer chromosomes have more protein coding genes. But that made me curious about variation.
```{r}
ggplot(genome, aes(x = protein_codinggenes, y = basepairs)) +
  geom_smooth(formula = y ~ x, method = lm, se = TRUE) +
  geom_point(color = "darkblue", size = 3) +
  labs(x = "Protine Coding Genes (#)", y = "Basepairs (Millions)", title = "Longer chromosomes have more protein coding genes in humans", subtitle = "Number of basepairs compared to protein coding genes for each chromosome", caption = "Data from Wikipedia | Plot from @matthewhirschey") +
  scale_y_continuous(labels = c(100, 150, 200, 250)) +
  geom_text_repel(aes(label = chromosome)) +
  theme_minimal() +
  NULL

#variation EDA
genome %>% mutate(percent_var = variations/basepairs) %>% select(chromosome, percent_var) %>% arrange(percent_var)

#Y chromosome has much* less variation. Let's see what this looks like...
```

## EDA Graph  
I chose to investigate the relationship between the number of basepairs and the amount of variation on each human genome. 
```{r}
ggplot(genome, aes(x = basepairs, y = variations)) +
  geom_point(size = 3) +
  geom_smooth(formula = y ~ x, method = lm) +
  #labs(x = "Number of Basepairs", y = "Number of Genomic Variants", title = "Genomic Variaion is lowest on the human X and Y chromosomes", subtitle = "Comparison ", caption = "Data from Wikipedia | Plot from @hirscheylab") +
  theme_minimal() +
  #scale_y_continuous(labels = c(50, 100, 150, 200, 250)) +
  geom_text_repel(aes(label = chromosome)) +
  NULL
```
  
**The X and Y chromosomes are clearly lower, but how to visualize this?**  
  
## Final Graph
I chose the lollipop plot method to visualize the difference between each chromosome's variation compared to the mean variation
```{r final_graph}
#autocalc mean variation for graph
mean_variation <- mean(genome$percent_var)

#this is a string that labels the graph using the annotate() layer
desc <- c("The XY Chromosomes have lower genomic varation than the others")

#xy var
x_var <- genome %>% filter(chromosome == "X") %>% select(percent_var) %>% pull()
y_var <- genome %>% filter(chromosome == "Y") %>% select(percent_var) %>% pull()

#this is code to draw a path (line) on the graph to match the style of ggforce()
path <- tibble(
  x = c(.03, .03, .029, mean_variation),
  y = c(18.5, 20.5, 21.5, 21.5)
)

#this is code to draw arrows
arrows <- tibble(
  x1 = c(.03, .01),
  x2 = c(.034, 0.006),
  y1 = c(4.5, 4),
  y2 = c(2, 2.5)
)

ggplot(genome, aes(x = percent_var, y = fct_rev(chromosome))) +
  geom_vline(aes(xintercept = mean_variation), color = "gray70", size = 0.6) +
  geom_segment(aes(y = fct_rev(chromosome), yend = fct_rev(chromosome), x = mean_variation, xend = percent_var), size = 0.4, color = "gray70") +
  geom_mark_ellipse(aes(filter = chromosome == "Y", label = 'Sex Chromosomes', 
                        description = desc, fill = xy)) +
  geom_mark_ellipse(aes(filter = chromosome == "X", fill = xy)) +
  geom_point(size = 3, color = "gray20") +
  theme_minimal() +
  labs(x = "Variation (percent)", y = "Chromosome", title = "Genomic variaion is lowest on the human X and Y chromosomes", subtitle = "Comparing number of variations to number of base pairs on human chromosomes", caption = "Data from Wikipedia | Plot from @matthewhirschey") +
  annotate("text", x = 0.03, y = 17, size = 4, color = "gray20",
           label = glue::glue("Variation average\n{round(mean_variation*100, 1)}%")) +
  geom_path(data = path, aes(x = x, y = y)) +
  theme(legend.position = "none") + #removes the lengend
  coord_cartesian(ylim = c(-0.15, 24)) +
  annotate("text", x = 0.03, y = 6, size = 3, color = "gray60",
           label = glue::glue("X variation\n{round(x_var*100, 1)}%")) +
  annotate("text", x = 0.014, y = 3.2, size = 3, color = "gray60",
           label = glue::glue("Y variation\n{round(y_var*100, 1)}%")) +
  geom_curve(data = arrows, aes(x = x1, y = y1, xend = x2, yend = y2),
             arrow = arrow(length = unit(0.07, "inch")), size = 0.4,
             color = "gray60", curvature = 0.2) +
  scale_x_continuous(labels = c(0, 2, 4, 6)) +
  NULL

#save your new plot to have an image for your final presentation
ggsave(here::here("5_markdown", "variation.png"), plot = last_plot(), height = 5, width = 7, units = "in", dpi = 300) #plan on changing the width:height ratio for your own plot
```

## Conclusions
The amount of genomic variation is lowest on the X and Y chromosomes, compared to the remaning human chromosomes. This observation implies that the fidelity of DNA sequences is more stringent and unique mechanisms are in place to ensure low variation.  
  
## Prioritized follow-up studies
Chromatin states are known to influence genomic variation. Therefore, I would like to look at the chromatin state at these genomic loci, including histone modifications and DNA methylation status.

## Acknolwedgements
I would like to acknowledge Cedric Scherer for [plot inspiration](https://cedricscherer.netlify.com/2019/05/17/the-evolution-of-a-ggplot-ep.-1/), @garretgrom, @hadleywickham, and the @rstuio team for making the tidyverse packages easy to use, and you all for reaching the end of the class!